Demo on how to emulate classical kernels

emgilfuster opened the pull request:
This demo introduces a re-uploading in depth quantum embedding kernel and demonstrates how it can approximate any stationary classical kernel up to arbitrary precision.

emgilfuster commented:
Hi @mariaschuld, would you like to have a look? :)

mariaschuld commented on demonstrations/tutorial_expressivity_fourier_series.py:
```diff
@@ -205,7 +205,7 @@ def target_function(x):
     res = coeff0
     for idx, coeff in enumerate(coeffs):
-        exponent = np.complex(0, scaling * (idx+1) * x)
+        exponent = np.complex128(scaling * (idx+1) * x * 1j)
```
why is this needed I wonder?

emgilfuster replied:
this is from the earlier PR, I'll get rid of it

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+============================
+
+.. meta::
+    :property="og:description": Show how to approximate the Gaussian kernel.
```
```suggestion
    :property="og:description": Approximating the Gaussian kernel with quantum circuits.
```
Just a suggestion...

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+In this demo we will briefly revisit some notions of kernel-based Machine
+Learning (ML) and introduce one very specific `Quantum Embedding Kernel (QEK)
+<https://pennylane.ai/qml/demos/tutorial_kernels_module.html>`_.
+We will also use this QEK to demonstrate we can approximate a large class of
```
```suggestion
We will use this QEK to demonstrate we can approximate a large class of
```

It's  kind of the main thing...

mariaschuld replied:
Maybe "important" class? Size is relative in function spaces...

emgilfuster replied:
yes, important is definitely more accurate

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+----------------------------------------------
+
+As we just said, in the interest of keeping the concepts at a high level we
+will not be reviewing all the notins of kernels in-depth here.
```
```suggestion
will not be reviewing all the notions of kernels in-depth here.
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+#. `Training and evaluating quantum kernels <https://pennylane.ai/qml/demos/tutorial_kernels_module.html>`_
+#. `Kernel-based training of quantum models with scikit-learn <https://pennylane.ai/qml/demos/tutorial_kernel_based_training.html>`_
+
+Now, a kernel for us is a real-valued function of two variables
```
```suggestion
Now, for the purpose of this demo, a kernel is a real-valued function of two variables
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+positions :math:`k(x_1,x_2) = k(x_2,x_1)`.
+Finally, we will also want to enforce the kernels be positive semi-definite,
+but let's avoit getting lost in mathematical definitions, you can trust that
+all kernels featuring in this demo are positive semi-definite.
```
I like this, it gives the reader the feeling she is looked after somehow :)

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+all kernels featuring in this demo are positive semi-definite.
+
+If you take whichever textbook on kernel methods and search for the word
+``prominent", chances are you'll find it next to the word ``example" in a
```
```suggestion
"prominent", chances are you'll find it next to the word "example" in a
```
This is not latex ;)

emgilfuster replied:
oopsie!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+If you take whichever textbook on kernel methods and search for the word
+``prominent", chances are you'll find it next to the word ``example" in a
+sentence that introduces the so-called Gaussian (Radial Basis Function) kernel
+:math:`k_\sigma`.
```
This required me to think for a while, but I had to laugh - you are so right!

emgilfuster replied:
if it's too confusing we could also take it out!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+If you take whichever textbook on kernel methods and search for the word
+``prominent", chances are you'll find it next to the word ``example" in a
+sentence that introduces the so-called Gaussian (Radial Basis Function) kernel
```
```suggestion
sentence that introduces the so-called Gaussian (or Radial Basis Function) kernel
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+.. math:: k_\sigma(x_1, x_2) = e^{-\frac{\lvert x_1 - x_2\rvert^2}{\sigma}},
+
+where the variance :math:`\sigma` is a positive real tunable parameter (the
+generalization to higher-dimensional data is straightforward using the
+Euclidean norm :math:`\lVert x_1 - x_2 \rVert_2^2`.)
```
```suggestion
where the variance :math:`\sigma` is a positive real tunable parameter. The
generalization to higher-dimensional data is straightforward using the
Euclidean norm :math:`\lVert x_1 - x_2 \rVert_2^2`.
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+generalization to higher-dimensional data is straightforward using the
+Euclidean norm :math:`\lVert x_1 - x_2 \rVert_2^2`.)
+Now for practical purposes the Gaussian kernel has the advantage of being
+simple enough to study as a function, while still yielding good performance for
+real-life tasks.
```
```suggestion
simple enough to study as a function, while yielding good performance for a wide range of
real-life tasks.
```
Actually, it is quite puzzling how good this simple kernel is - we don't only use it for ease of analysis! Also, a Gaussian kernel can be shown to be universal, which means that the RKHS of that kernel contains _all_ functions.

emgilfuster replied:
I like your suggestion, sounds much better!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+real-life tasks.
+
+In particular, one of the properties of the Gaussian kernel is that it is a
+shift-invariant function (also called stationary).
```
```suggestion
shift-invariant (also called *stationary*) function .
```

(I think this is the italics in rst, but always mix them up)

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+.. math:: k_\sigma(x_1, x_2) = e^{-\frac{\lvert x_1 - x_2 \rvert^2}{\sigma}} =e^{-\frac{\delta^2}{\sigma}} = k_\sigma(\delta).
+
+Also, this combined with the property :math:`k(x_1, x_2) = k(x_2, x_1)` results
```
```suggestion
Combined with the property :math:`k(x_1, x_2) = k(x_2, x_1)`, this results
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+in the new property :math:`k(\delta)=k(-\delta)`.
+
+Of course the Gaussian kenrel is not the only shift-invariant kernel out there.
+As it turns out, there are many others (REFERENCES OR POINTERS OR SOMETHING!)
```
You could link to the Rasmussen textbook, I think it is free for download!

emgilfuster replied:
sorry, this slipped past me, I was supposed to have the link here, not the reminder to add a link! That's a good idea though

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+Nevertheless, here we will only look at the simple Gaussian kernel with
+:math:`\sigma = 1`:
+
+.. math:: k_1(\delta) = e^{-\delta^2},
+
+but all the arguments and code we use are also amenable to other kernels which
+fulfill the following mild restrictions:
+#. Shift-invariance.
+#. Normalization :math:`k(0)=1`.
+#. Smoothness (seen as quickly decaying Fourier spectrum).
```
```suggestion
.. math:: k_1(\delta) = e^{-\delta^2}.

All the arguments and code we use are also amenable to other kernels which
fulfill the following mild restrictions:

* Shift-invariance.
* Normalization :math:`k(0)=1`.
* Smoothness (seen as quickly decaying Fourier spectrum).
```

The hashtags don't seem to work?

emgilfuster replied:
the problem was that I didn't leave an empty line between the paragraph and the list

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+#. Normalization :math:`k(0)=1`.
+#. Smoothness (seen as quickly decaying Fourier spectrum).
+
+Let's warm up our coding by writing the simple Gaussian kernel!
```
```suggestion
Let's warm up by implementing a classical Gaussian kernel!
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+Let's warm up our coding by writing the simple Gaussian kernel!
+
+First, importers gonna import B-)
```
```suggestion
First, importers gonna import B-).
```

I am not sure I get this one, but that should not be a criterion :)

emgilfuster replied:
It's an adaption from "haters gonna hate", but it's probably one too many, I'll leave this comment open and rethink in next iteration whether we keep it or discard it

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+import pennylane as qml
+from pennylane import numpy as np
+import matplotlib.pyplot as plt
+import math 
+
```
```suggestion
import math 
```

emgilfuster replied:
There is no suggested change, is that right?
Probably I could get away with only numpy, no need for math?

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+np.random.seed(53173)
+
+###############################################################################
+# now we define the function and plot it as a mild sanity check:
```
```suggestion
# Now we define the kernel function and plot it as a mild sanity check:
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+plt.plot(X, Y_gaussian)
+plt.suptitle("The gaussian kernel with $\sigma=1$")
+plt.show();
```
I think we need a x axis label here - people plot kernels usually by fixing one argument and varying the other, but since things are stationary, this is the same as varying the delta. So good to be explicit

emgilfuster replied:
yes, true!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+Also, this combined with the property :math:`k(x_1, x_2) = k(x_2, x_1)` results
+in the new property :math:`k(\delta)=k(-\delta)`.
+
+Of course the Gaussian kenrel is not the only shift-invariant kernel out there.
```
```suggestion
Of course the Gaussian kernel is not the only shift-invariant kernel out there.
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+###############################################################################
+# That's exactly what I have in mind when I think of a Gauss bell curve.
+# So far so good!
```
About the above section: Even if it's short a few subtitles may work well. It's like a magazine vs a long newspaper article - people keep reading. For example "The Gaussian kernel", "Shift-invariance", and ... hm.. maybe "Visualisation"? Ok, the last one is not too great

emgilfuster replied:
ok I'll try that out

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# --------------------------
+#
+# The quantum algorithms we build today belong to the family of Quantum
+# Embedding Kernels (QEKs), introduced in the paper [#QEK]_ corresponding to
```
I know someone who may jump on your roof when you make this public :) It sounds like quantum kernels were invented in the paper. To be safe, one could be more specific and say that the name was introduced there, after others proposed the idea of quantum kernels?

emgilfuster replied:
yeah the wording is quite unfortunate, my bad! I only wanted to mention the paper as a more comprehensive explanation than the demo. But actually the demo is already quite comprehensive, so maybe we don't even need it. I'll try reshuffling things and see what works best!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# 
+# Now, oftentimes the feature map will be nothing other than applying a unitary
+# gate depending on the data :math:`U(x)` to the ground state of the
+# qu:math`d`it:
```
Not rendering

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# And this is the bare bone of what we call Quantum Embedding Kernels!
+#
+# By now you may have already realized QEKs are very cool and all but they have
+# one shortcoming: in general :math:`k_Q` won't be a shift-invariant function
```
```suggestion
# one shortcoming: In general, :math:`k_Q` won't be a shift-invariant function
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+#
+# By now you may have already realized QEKs are very cool and all but they have
+# one shortcoming: in general :math:`k_Q` won't be a shift-invariant function
+# (but is this really a shortcoming?)
```
```suggestion
# (but is this really a shortcoming?).
```

I think this is too contrived, I would just state that they have a shortcoming when our goal is to emulate the Gaussian kernel on a quantum computer (and it may be a good opportunity to repeat what we're trying to do here).

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# (but is this really a shortcoming?)
+# The key word here is *general*.
+# General QEKs are not shift-invariant, so one reasonable question would be:
+# which restrictions do we have to impose for these kernels to be
```
```suggestion
# Which restrictions do we have to impose for these kernels to be
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# first step towards it.
+# Namely, we next introduce one particular class of QEKs which are indeed
+# shift-invariant.
+# Behold, ladies and gents, *the Boring kernel*!
```
Lol, you really gonna put this name out into the wild?

I fear this won't make external review :) The reason is not that one cannot use humorous names, but that it is not a very useful description of what we are doing. Is there any alternative? 

In my mind, what this is is a toy model! (And if it turns out to be a good one, it may keep thousands of physicists of future generations restless...)

emgilfuster replied:
hehe will change

emgilfuster replied:
I called it "the stationary toy QEK"

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# The Boring kernel
+# ------------------
+#
+# The boring kerne is a stationary kernel that can be estimated with a
```
```suggestion
# The boring kernel is a stationary kernel that can be estimated with a
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# the trouble of finding a good measurement observable.
+# Indeed, for a given feature map :math:`\rho:\mathcal{X}\to\mathcal{H}_d`,
+# where :math:`\mathcal{H}_d` is the Hilbert space of a quantum system of
+# :math:`d` dimensions (also called a qu-:math:`d`-it), we have the associated
```
Some readers may be worrying "but my Master thesis is about qubits not qudits, how can I use this?" I think we need a sentence that explains how this can be thought of as multiple qubits, or just go the qubit route from the start!

emgilfuster replied:
Since I argue everything in terms of qubits below, should I just add a disclaimer saying that noone should worry about qudits, because when it comes to implementation we do everything with qubits? 

Or rather add a sentence building a bridge between one qudit and several qubits and keep the other things as they are?

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# input data.
+# The data is encoded as the time parameter in the time evolution of a diagonal
+# number-operator hamiltonian :math:`H_d=\operatorname{diag}(0, 1, \ldots,
+# d-1)`.
```
We definitely need a circuit drawing here, this is confusing even me :)

emgilfuster replied:
right!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# we reach an expression for the boring
+# kernel :math:`k_d`:
+#
+# .. math:: k_d(\delta) = \sum_{s=-(d-1)}^{d-1} a_s e^{i\delta s},
```
Hmm, I agree that we cannot roll out the theory here, but this innocent line is only obvious to a handful of people in the universe. Maybe we can at least mention the flavour, i.e., "we know that many quantum kernels, ours included, can be expressed as Fourier series of only a few terms". You can also cite my kernels paper, I think this is the first (and only?) reference that has an explicit Fourier expression of general quantum kernels...

emgilfuster replied:
yip, I'll do that

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# --------------------------------------------
+#
+# Now that we've laid out the formulas, we only need to write down the PL code
+# that relized the quantum circuit, right?
```
```suggestion
# that realized the quantum circuit, right?
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# :math:`W` is a trainable :math:`d`-dimensional unitary, independent of the
+# input data.
+# The data is encoded as the time parameter in the time evolution of a diagonal
+# number-operator hamiltonian :math:`H_d=\operatorname{diag}(0, 1, \ldots,
```
I don't think many people know what this is, so maybe say "to make things simple, we need a Hamiltonian with consecutive integers on its diagonal; this is known in continuous-variable quantum computing as a number operator". And then move the explanation of how to do it in qubits to here? 

The goal is that people know why we do this, and that we can do this easily with the tools they know.

emgilfuster replied:
yes, very good call!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# spectrum :math:`(0, 1, \ldots, 2^n-1)`.
+#
+# We will also be spared of some details here (or, as your quantum mechanics
+# prof would say, the derivation is left as an exercise for the reader), but
```
```suggestion
# Prof would say, the derivation is left as an exercise for the reader), but
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# prof would say, the derivation is left as an exercise for the reader), but
+# one valid choice for this to happen is :math:`\vartheta_j = -2^{j-1}`.
+#
+# With this, we can *finally* start getting our hands dirty!
```
A slow reader will have spent half an hour getting here. Maybe sprinkling the text with reminders of the goal (to approximate a Gaussian kernel with a quantum circuit) could help to keep the thread going?

emgilfuster replied:
yes sire

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+###############################################################################
+# And we also provide a function that generates a random tensor of parameters
+# with the cirrect size.
```
```suggestion
# with the correct size.
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+###############################################################################
+# And finally we also define a function that gives us the vector of parameters
+# we need to feed to :math:`S(x)`
```
```suggestion
# we need to feed to :math:`S(x)`:
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# Computing the QEK on the quantum computer
+# ------------------------------------------
+#
+# That is to say, on the ``default.qubit`` PL quantum simulator ;)
```
```suggestion
# That is to say, on the ``default.qubit`` PL quantum simulator ;).
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+###############################################################################
+# Next we define the circuit function, which outputs the vector of
+# probabilities for the computational basis states.
+# This is also the point where we need the vector of ``thetas``
```
```suggestion
# This is also the point where we need the vector of ``thetas``:
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+
+###############################################################################
+# But we only need to keep the probability of the state
+# :math:`\vert00000\rangle` being measured:
```
A lot of students find this totally opaque. Maybe here (and before) you can point them towards places where they can read up on why this circuit evaluates a kernel, and what the alternatives are!

Actually, it would be so helpful to have a demo that just introduces the different circuits one can use to compute quantum kernels!

emgilfuster replied:
agree, and agree!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+print("for x1 = 0.1 and x2 = 0.6 is\n\tk(x1, x2) =", boring_kernel(.1, .6, thetas, random_pars))
+
+###############################################################################
+# Quality of life improvement: let's make this small function that evaluates
```
```suggestion
# Quality of life improvement: Let's make this small function that evaluates
```

mariaschuld replied:
Check the grammar, something is not right?

emgilfuster replied:
I think it's a confusing use of "this", I was trying to go for the use in oral storytelling "there was this guy, who...". But it's probably too confusing here, will change!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,491 @@
+# This was how you can approximate the Gaussian kernel using a boiled down QEK
+# we called the boring kernel!
+#
+# .. figure:: ../demonstrations/classical_kernels/salesman.PNG
```
Ah, there you got the circuit picture you can use above!

emgilfuster replied:
right! it's also in the thumbnail so I already have it ready

mariaschuld reviewed (changes requested):
A well-written and entertaining demo. When introducing humour into professional spaces one must be absolutely on point (else it is not taken serious), so we may need a few more revisions when other reviewers leave comments.

I have a few suggestions: 

1) Can we crystallise the main message even more throughout the demo, i.e., we want to approximate a Gaussian kernel using a quantum computer. [To give an example: when you plot the Gaussian kernel, just mention that this is what we want to do on a quantum computer.]
2) Sometimes you introduce small innocent bits that require years of study to understand, for example the idea of d-dits and number operators, or the idea of a Fourier representation of a quantum kernel. You don't need to go into depths, but acknowledge that this is something people need to swallow and give them references to look it up if needed.
3) I am in two minds about this one, but the real cool thing you do is totally swept under the rug (I guess to keep it simple): You construct a quantum kernel whose spectrum we can control [at least asymptotically], and use this to design kernels - in our example a classical kernel. This is an extremely powerful idea! What about the following: make this the main idea of the tutorial ("designing quantum kernels by shaping their Fourier spectrum") and sell the Gaussian kernel as a mere example! I fear though that one may have to move some sections around, but it is worth it I think! The demo could become a "build your own kernel on a quantum computer" sandbox.
4) The big weakness of the demo is the part when things are learnt - if we are trying to sell a simple toy model for playing, but then we need the huge apparatus of optimisation to get the final result, there is a certain anti-climax. I would at least explain this, namely that we can shape the right spectrum, but it is hard to guess which circuit leads to the right coeffs. 

PS: There is NO way we could analytically find the gates needed to form the coeffs, right? Even if we do something as rough as we did with the number operator, and construct the right W, and then use compilation to find the right gates?

emgilfuster commented:
Hi @mariaschuld, the demo is ready for code review after the major rewriting.

I still don't have the option to assign reviewers myself, sorry I'm tagging you this way.

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
I think it is important to explain here what you are actually doing. I.e. you could say that we could of course just take the function implementing the classical kernel and the function implementing the quantum kernel and find the amplitudes so that they match. But we want to go a different route here and derive an equation that links the amplitudes to the spectrum we want to construct, and then use old-fashioned convex optimisation to find the solution...

That reminds me, do you mention anywhere that this design of your circuit is cool because we could approximate any stationary kernel? Is that even true?

emgilfuster replied:
I said towards the very beginning that it would work for all kernels that fulfill the three conditions.
It might be too far above for anyone to remember, so should I drop a reminder somewhere around here?

emgilfuster replied:
I added a paragraph along the lines of what you proposed

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```suggestion
# These closed-form equations allow us to find the solution numerically, using
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
Ah here you are going first person again. I have no opinion on this style, I wonder if the others will comment on it though. Let's see?

emgilfuster replied:
I tuned it down, that was again too much

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```suggestion
# Good coding practice tells us we should probably turn this step into a function
```

I am not sure though :)

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
Again, I am not sure the reader understands how general this ansatz is!

emgilfuster replied:
Do you identify an optimal spot to repeat the message at the start of the demo? :)

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
:)

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
yes, that sounds very good!

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
yes, I like it!

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
missing a "We"

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
yep, better!

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
^^

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
I added more explanation to this. I'm not sure whether it's a critical point, so maybe it would be better to remove the entire thing altogether? I think an external look would be very helpful here

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
I added a couple lines here, let me know what you think

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
yeah, I guess it wasn't the right place for the distinction since I haven't talked enough about the distinction yet. Do you think I should have another visual schematic later on that stretches the two very different parts?

mariaschuld replied:
I think it's not necessary.

In fact, this image is so spot on  that I wonder if it should appear as an icon and as the first picture of the demo...it is much more specific than the one showing the circuit!

emgilfuster replied:
I did the swap, do the captions now sound misplaced? I'd say no, but wanted to ask

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
hm, maybe now it's better

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
agree! Also later, too hard change of style

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
good question, maybe I should!

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
I tried to rewrite it, does it sound less confusing now? I'm not yet 100% sure I'm happy with it, I might change it a bit when I do a final end-to-end read.

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
yep agree!

emgilfuster commented:
Hi @mariaschuld, I implemented the proposed, see unresolved conversations for the places where I wasn't sure.
I replaced all "QEK" mentions for "QK", also in the code

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+(PQCs) to find good functions, whatever *good* means here.
+Consequently, since kernels are just one specific kind of well-defined
+functions, the task of finding a quantum kernel (QK) that approximates a given
+classical one could be posed as a supervised learning regression problem.
```
```suggestion
classical one could be posed as an optimisation problem.
```

Would you mind this change? Learning is one option to match kernels via optimisation, but this is more general...

emgilfuster replied:
yes makes sense

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+The way to attack this task would be to define a loss function directly
+quantifying the distance between both functions (the classical kernel function,
+and the PQC-based hypothesis).
+This sort of approach is more or less oblivious to the Fourier representation
+of the kernels, though.
```
```suggestion
This sort of approach does not help us much to gain theoretical insights about the 
structure of kernel-reproducing quantum circuits, though.
```

As a reader I wondered: Why on earth should I care about Fourier? :) 
Maybe this is a bit more general - we want to be more theoretical, and in the next sentence you 
say which approach is chosen...

emgilfuster replied:
no, you're totally right! I think this was here from an earlier version where I had already mentioned our desire to explore the Fourier side of things. Off it goes!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+This sort of approach is more or less oblivious to the Fourier representation
+of the kernels, though.
+
+We want to study the link between classical and quantum kernels through the
+lens of the Fourier representation of a kernel, which is a common tool in
+classical ML.
```
```suggestion
In order to build intuition, we will instead study the link between classical and quantum kernels through the
lens of the Fourier representation of a kernel, which is a common tool in
classical ML.
```

Does that link it well?

emgilfuster replied:
yes I think so!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+In order to keep the demo short and sweet, we focus on one simple example, but
+the same ideas would apply for more general scenarios.
+Also, Refs. [#QEK]_, [#Fourier]_, and [#qkernels]_ should be helpful for those
+who'd like to see the underlying theory of QKs (and also Quantum Embedding
+Kernels) and their Fourier representation.
```
```suggestion
who'd like to see the underlying theory of QKs (and also so-called _Quantum Embedding
Kernels_) and their Fourier representation.
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+
+.. figure:: ../demonstrations/classical_kernels/QEK.jpg
+    :align: center
+    :width: 80%
```
```suggestion
    :width: 70%
```
Or even smaller? I always like pictures to be scaled such that the text they contain is of the same size as the normal text... [Minor one]

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+:math:`k(x_1,x_2) = k(x_2,x_1)`.
+We also enforce kernels to be positive semi-definite, but let's avoid getting
+lost in mathematical lingo, you can trust that all kernels featuring in this
+demo are positive semi-definite.
```
I really like how this section abstracts away some details and refers to other resources!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+Start importing the usual suspects:
+"""
+
+import pennylane as qml
+from pennylane import numpy as np
+import matplotlib.pyplot as plt
+import math
+
+
```
```suggestion
import pennylane as qml
from pennylane import numpy as np
import matplotlib.pyplot as plt
import math

```

mariaschuld replied:
```suggestion

```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+def make_data(n_samples, lower=-np.pi, higher=np.pi):
+    x = np.linspace(lower, higher, n_samples)
+    y = np.array([gaussian_kernel(x_) for x_ in x])
+
```
```suggestion
```

A matter of taste :)

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+X, Y_gaussian = make_data(100)
+
+plt.plot(X, Y_gaussian)
+plt.suptitle("The gaussian kernel with $\sigma=1/\sqrt{2}$")
```
```suggestion
plt.suptitle("The Gaussian kernel with $\sigma=1/\sqrt{2}$")
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# Fourier analysis of the Gaussian kernel
+# -------------------------------------------
+#
+# The first next step will be to find the Fourier spectrum of the Gaussian
```
```suggestion
# The next step will be to find the Fourier spectrum of the Gaussian
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+#
+# The first next step will be to find the Fourier spectrum of the Gaussian
+# kernel, which is an easy problem for classical computers.
+# Once we've found it, we'll build a QK that produces a finite Fourier series
```
Just a thought: is there any reason to use an abbreviation? Doesn't it trade clarity with the advantage of saving a few characters? :) Even I had to think what QK meant here...

emgilfuster replied:
hm fair point. I usually like abbreviations where there are paragraphs that just repeat "quantum kernels" all the times. But glancing the document up and down I couldn't find any such instance. That being said, I think I'd leave it as is?

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# for each frequency :math:`n`.
+# The real part corresponds to the :math:`a_n` coefficient, and the imaginary
+# part to the :math:`b_n` coefficient: :math:`c_n=a_n+ib_n`.
+# Because the Gaussian kernel is an even function, we know the imaginary part
```
```suggestion
# Because the Gaussian kernel is an even function, we know that the imaginary part
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+
+def fourier_p(d):
+    # we only take the first d coefficients [:d]
+    # because `coefficients` treats the negative frequencies
```
```suggestion
    # because coefficients() treats the negative frequencies
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# of every coefficient will be zero, so :math:`c_n=a_n`.
+
+def fourier_p(d):
+    # we only take the first d coefficients [:d]
```
```suggestion
    # We only take the first d coefficients [:d]
```

If it has a fullstop the first word is capitalised...

emgilfuster replied:
correcto! That one slipped past me

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# terms.
+# But isn't that problematic, one may say?
+# Well, maybe.
+# Since we know the Gaussian kernel is a smooth function, we expect that the
+# coefficients converge to :math:`0` at some point.
+# We will not need to take more terms than this point, since all the extra ones
+# would have to be :math:`0` in any case.
```
```suggestion
# Since we know the Gaussian kernel is a smooth function, we expect that the
# coefficients converge to :math:`0` at some point, and we will only need to 
# consider terms up to this point.
```

Minor one, but I think the argument was repeated here!

emgilfuster replied:
agree, I was splitting hairs when I wrote it down, and it doesn't make much sense

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# The quantum embedding circuit will consist of two parts.
+# The first one, trainable, will be a parametrised general state preparation
+# scheme :math:`W_a`, with parameters :math:`a`. 
+# In the second one we input the data, and call it :math:`S(x)`.
```
...and the image that is currently in the first section would be an excellent fit here, because you explain what the 16x, 8x... actually mean!

emgilfuster replied:
a-ha! that's a good one, thanks!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+#   S(\delta)W_a\lvert0\rangle\rvert^2.
+#
+# So far we only wrote the gate layout for the quantum circuit, no measurement!
+# We need a few more functions for that!
```
Very nicely explained this section!

emgilfuster replied:
very kind, and thanks for the patience on your side!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+plt.xlabel("frequency $n$")
+plt.ylabel("Fourier coefficient $c_n$")
+plt.suptitle("Fourier spectrum of the Gaussian kernel")
+plt.show();
```
This section is very well explained. I think some people will get distracted by the detail of how one determines how many coefficients are needed. But the last plot should bring them back...

emgilfuster replied:
I had my doubts here. Maybe it wasn't necessary to go into this level of detail but I didn't want people to think the cut-off came from under a stone or was arbitrary.

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+dev = qml.device("default.qubit", wires = n_wires, shots = None)
+
+###############################################################################
+# And write the function that will use the device, hence the ``@qml.qnode``
+# decorator
```
```suggestion
# Next we construct the quantum node:
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+###############################################################################
+# Recall that the outputs of QKs can be defined as the probability of obtaining
+# the outcome :math:`\lvert0\rangle` when measuring on the computational basis.
+# That corresponds to the :math:`0^\text{th}` entry of ``qml.probs``:
```
Nicely explained!

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# The necessary little bit of math
+# ----------------------------------
+#
+# One could just take the function implementing the classical kernel and the
+# function implementing the quantum kernel and then tune the parameters of the
+# quantum kernel in a gradient-based manner until it matched the classical one.
+# But we want to go a different route here.
+# For that, we derive an equation that links the amplitudes to the spectrum we
+# want to construct, and then use old-fashioned convex optimisation to find the
+# solution.
```
```suggestion
# If we knew exactly which amplitudes to choose in order to build a given Fourier spectrum,
# our job would be done here. However, the equations derived in the literature are not 
# trivial to solve.
#
# As mentioned in the introduction, one could just "learn" this relation: 
# take the function implementing the classical kernel and the
# function implementing the quantum kernel and then tune the parameters of the
# quantum kernel in a gradient-based manner until it matches the classical one.
#
# We want to take an intermediate route between analytical solution and black-box optimisation here.
# For that, we derive an equation that links the amplitudes to the spectrum we
# want to construct, and then use old-fashioned convex optimisation to find the
# solution.
```

mariaschuld replied:
What about this one?

mariaschuld replied:
Actually, why not making this its own subsection under the header "How to find the amplitudes emulating a Gaussian kernel" or so. And then a reader can jump the next few sections if necessary, which most of them will probably do at first.

emgilfuster replied:
ok so I changed the title from "the necessary bit of math" into "How to find the amplitudes emulating a Gaussian kernel" and demoted it from being a section to being a subsection of the previous one "Computing the QK function on a quantum device"

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+                J[i][j] += probabilities[i + j]
+            if(i - j <= 0):
+                J[i][j] += probabilities[j - i]
+
```
```suggestion
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# For Newton's method we also need an initial guess.
+# Finding a good initial guess requires some tinkering, different problems will
+# benefit from different ones.
+# Here is a tame one that works for the gaussian kernel.
```
```suggestion
# Here is a tame one that works for the Gaussian kernel.
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+spectrum = fourier_p(2 ** n_wires)
+
+###############################################################################
+# Fix the hyperparameters for Newton's method
```
```suggestion
# We fix the hyperparameters for Newton's method:
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+# of them must've turned negative.
+# This isn't necessarily bad in general, only for us.
+# For ``MottonenStatePreparation`` we'll need to give ``amplitudes`` as one of the
+# arguments, which is the componentwise square root of ``probabilities`` so it
```
```suggestion
# arguments, which is the component-wise square root of ``probabilities`` so it
```

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+plt.plot(range(d), amplitudes, 'x', label = "amplitude $a_j$")
+plt.xlabel("array entry $j$")
+plt.legend()
+plt.show();
```
Hm I wonder if you could shorten this thresholding, because it is a small technical detail that can be quite distracting...took me a while to understand that you are just cleaning the results a bit to not run into numerical problems with the square root.

emgilfuster replied:
hm the code is necessary, otherwise we get nan in the vector of amplitudes in a lot of places. What I could do is cut the text? But then again I fear it would feel like it's coming from under a stone, and maybe a bit arbitrary. What do you think? For now I don't touch anything here.

mariaschuld replied:
Ok convinced, no worries unless another reviewer brings it up :)

mariaschuld commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,738 @@
+plt.xlabel("$\delta$")
+plt.ylabel("$k(\delta)$")
+plt.legend()
+plt.show();
```
Very cool Finale!

mariaschuld reviewed (changes requested):
Hey @emgilfuster. Great job. From the outside it looks like the same number of comments, but they are a lot lighter, and all of them are just suggestions or layout practices that other demos adopted...so it is just that extra cherry on top.

The only two serious ones are also minor:
1. consider swapping the 4 circles graphic with the QEK one
2. shorten the "thresholding probablilites" part to hide it as much as you can

I hope it'll be very quick to implement these, and then we open for other reviewers - but I think it's in a very good shape to be accepted soon.

Thanks for working on it!

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+# yet like a Gaussian.
+#
+# How to find the amplitudes emulating a Gaussian kernel
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I changed the title from "the necessary bit of math" to this, and demoted it from being a standalone section to being a subsection of "Computing the QK function on a quantum device", is this more or less what you had in mind?

mariaschuld replied:
Sounds good!

emgilfuster commented:
Hi @mariaschuld thank you very much for the on-point feedback!
I think I implemented everything except for the thresholding thing, which I don't really know how to do further. I considered turning it into a note or a warning, but it's really necessary for the core of the code, so I didn't do it in the end. In the text I dress it as "good coding practise", but the truth is that if I don't include the thresholding everything breaks, because some of the amplitudes that come from Newton's method are very small but negative, so taking the square root just kills us.

Do you think it can stay as it is? Should I add a disclaimer about skippability?

mariaschuld reviewed (approved):
Thanks for making the changes @emgilfuster ! I will tag external reviewers, let's see what they think of the stuff we cooked up...

peterse commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+#
+# In order to simplify the formulas, we introduce new variables, which we call
+# ``probabilities`` :math:`(p_0, p_1, p_2, \ldots, p_{2^n-1})`, and we define as
+# :math:`p_j=a_j^2`.
```
nit: |a_j|^2

peterse commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+# ``MottonenStatePreparation`` as the *amplitudes* of a quantum state?
+# Then it makes sense that we call its squares *probabilities*, doesn't it?
+#
+# There is a crazy formula that matches the entries of *probabilities* with the
```
I'm curious about this formula. Where does it come from? Does this just fall out of the specific choices of W, S made for this kernel?

peterse replied:
this seems especially relevant considering the closing meme :p - this choice is general to _any_ translation invariant kernel then?

emgilfuster replied:
It follows directly from S, and how the entries of W relate to each Fourier frequency!
I added three lines on this :)

peterse commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+    return predict_spectrum(probabilities)[:d] - spectrum[:d]
+
+###############################################################################
+# These closed-form equations allow us to find the solution numerically, using
```
for readers unfamiliar with convex optimization it might be helpful to mention why this function is guaranteed to be convex?

emgilfuster replied:
yes, I agree! 
I added two sentences on this

peterse commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+
+plt.plot(range(d), probabilities, 'x')
+plt.xlabel("array entry $j$")
+plt.ylabel("probabilities $w_j$")
```
maybe be a bit more specific about which probabilities these are? are these our estimates for the things we were calling $p_j$ earlier?

emgilfuster replied:
uh, precisely! this slipped past me, thank you!

peterse reviewed (commented):
very cool. I made some minor comments and think it could use a bit of clarification in places but overall seems good to go.

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+for n in range(2,7):
+    N.append(n)
+    F = fourier_p(n)
+    plt.plot(N, F, 'x', label='{}'.format(n))
```
I'm no expert myself on the fast fourier transform, which is what's being called under the hood here.
But when I read a bit, apparently one requirement is to have a good frequency cut-off, otherwise the coefficients have some artifacts, which is what we're seeing here.
I think the reason is that specifying the cut-off specifies a set of functions with which you want to approximate a target function.
If I want to approximate cos(4x) as a linear combination of {1, cos(x), cos(2x), cos(3x), cos(4x)}, then of course I'll find coefficients (0, 0, 0, 0, 1).
Yet, if I tried to approximate it as a linear combination of {1, cos(x), cos(2x)}, I'd probably be able to find some coefficients (a, b, c)!=(0, 0, 0) that achieve a lower approximation error.

Exact same thing happens here, the Gaussian kernel "needs" at least up to {cos(6x)}, so anywhere below that we get artifacts from poor choice of hyperparameters at the FFT level.

emgilfuster commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,742 @@
+
+###############################################################################
+# Next we introduce the only trainable gate we neet to make use of.
+# Contrary to the usual Ansätze used in supervised and unsupervised learning,
+# we use a state preparation template called ``MottonenStatePreparation``.
+# The unitary associated to this template transforms the :math:`\lvert0\rangle`
+# state into a state with amplitudes :math:`a=(a_1,a_2,\ldots,a_{2^n-1})`,
```
I added two sentences regarding the confusion.
I changed a_1 into a_0

co9olguy commented on demos_qml.rst:
```diff
@@ -92,6 +92,12 @@ IBM, Google, Rigetti, and many more—or on real hardware devices.
 
+.. customgalleryitem::
+    :tooltip: Estimate a classical kernel function on a quantum computer.
+    :figure: demonstrations/classical_kernels/gaussian_kernel.png
```
The picture with the circuit (or even the meme!) would make a more eye-catching thumbnail here

emgilfuster replied:
hm, I thought the text in the meme would become too small to be understood? And similar with the other one, wouldn't it appear like any other circuit? Plus wouldn't the aspect ratio be a problem? 
I'm asking because you have more experience here, ofc I'll go for the most blatant clickbait!

mariaschuld replied:
Thanks, I overlooked this...I would put the 4 circle graphic?

emgilfuster replied:
right away! chop chop

emgilfuster commented:
Hi @peterse I think I implemented all your proposals.
I left some answers to your comments, feel free to resolve them if you agree with what I wrote :)

Thank you very much for the review! 
I found it very spot on and agreed with everything you said.

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+In order to keep the demo short and sweet, we focus on one simple example, but
+the same ideas would apply for more general scenarios.
+Also, Refs. [#QEK]_, [#Fourier]_, and [#qkernels]_ should be helpful for those
+who'd like to see the underlying theory of QKs (and also so-called *Quantum Embedding
```
*Quantum Embedding Kernels* not rendering well for me, for some reason.

emgilfuster replied:
there was an earlier version where the itallics was written with underscores instead of asterisks, and that didn't render for me either. In theory it should be fixed now, does the problem persist your end?

alvaro-at-xanadu replied:
This is fixed now :)

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+
+For the purposes of this demo, a *kernel* is a real-valued function of two
+variables :math:`k(\cdot,\cdot)` from a given data domain :math:`x_1,
+x_2\in\mathcal{X}`.
```
Is \mathcal{X} a subset of the real numbers? It can be an abstract set, but I guess we're restricting to something less abstract in this tutorial?

emgilfuster replied:
I added a line mentioning vector spaces, so that we avoid the problem you mention later.
Would you say this is enough?

Thanks for pointing this one out, indeed at no point did I introduce the extra structure required for shift invariant kernels to make sense, nice catch!

alvaro-at-xanadu replied:
This will do I think!

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+:math:`k(x_1,x_2) = k(x_2,x_1)`.
+We also enforce kernels to be positive semi-definite, but let's avoid getting
+lost in mathematical lingo, you can trust that all kernels featuring in this
+demo are positive semi-definite.
```
I agree that not much mathematical abstraction is needed, and I think this is great! However positive-semidefiniteness can be defined in one line and is not a hard concept, and since it's already mentioned I think it's worth to define.

emgilfuster replied:
hm, positive semidefiniteness for matrices can be defined in one line, I agree.
But PSDness for functions requires a bit more, wouldn't you agree?
Either you need to introduce the Gramm matrix, or you spell out Mercer's condition.

It's still not that much, but given that it doesn't add much content, maybe I'd leave it out

alvaro-at-xanadu replied:
Ah you're right, I thought it was easier! Better leave as is!

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+Shift-invariant kernels are those whose value doesn't change if we add a shift
+to both inputs. That means, for any suitable shift vector
+:math:`\zeta\in\mathcal{X}`, shift-invariant kernels are those for which
+:math:`k(x_1+\zeta,x_2+\zeta)=k(x_1,x_2)` holds.
```
Here is where it is implied that \mathcal{X} should be a vector space (this shift-rule doesn't make sense, for example, if \mathcal{X} does not contain a zero element.

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+
+Some kernels fulfill another important restriction, called *shift-invariance*.
+Shift-invariant kernels are those whose value doesn't change if we add a shift
+to both inputs. That means, for any suitable shift vector
```
```suggestion
to both inputs. Explicitly, for any suitable shift vector
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+to both inputs. That means, for any suitable shift vector
+:math:`\zeta\in\mathcal{X}`, shift-invariant kernels are those for which
+:math:`k(x_1+\zeta,x_2+\zeta)=k(x_1,x_2)` holds.
+At the same time, having this property means the function can be written in
```
```suggestion
Having this property means the function can be written in
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+:math:`k(x_1+\zeta,x_2+\zeta)=k(x_1,x_2)` holds.
+At the same time, having this property means the function can be written in
+terms of only one variable, which we call the *lag vector*
+:math:`\delta:=x_1-x_2\in\mathcal{X}`, abusing notation a bit:
```
```suggestion
:math:`\delta:=x_1-x_2\in\mathcal{X}`.  Abusing notation a bit:
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+# periodic function using the sine and cosine functions.
+# Fourier analysis tells us that we can write any given periodic function as
+#
+# .. math:: f(x) = a_0 + \sum_n a_n\cos(n\omega_0x) + b_n\sin(n\omega_0x).
```
Sums have to range over a set (1 to infinity in this case)

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+
+##################################################################################
+# In practice we'll be interested in cases where the period approaches
+# infinity.
```
Which means that it's not periodic anymore, this confuses me a bit.

emgilfuster replied:
I paraphrased the text here, hope it's more understandable now

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,754 @@
+    # We only take the first d coefficients [:d]
+    # because coefficients() treats the negative frequencies
+    # as different from the positive ones.
+    # For real functions, they are the same.
```
You can comment using """ (yes even in this markdown language) It looks better than writing a lot of #

emgilfuster commented:
Hi @alvaro-at-xanadu thanks for the review.
I implemented most of your suggestions and resolved them.
Would you check the ones I didn't resolve, and see if you're happy with what I wrote?

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+#
+# Designing a suitable QK amounts to designing a suitable parametrized quantum
+# circuit.
+# Let's take a moment to refresh the big scheme of things, with the following
```
```suggestion
# Let's take a moment to refresh the big scheme of things with the following
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# We construct the quantum kernel from a quantum embedding (see the demo on
+# `Quantum Embedding Kernels <pennylane.ai/qml/demos/tutorial_kernels_module.html>`_).
+# The quantum embedding circuit will consist of two parts.
+# The first one, trainable, will be a parametrised general state preparation
```
```suggestion
# The first one, trainable, will be a parametrized general state preparation
```

alvaro-at-xanadu replied:
Since Xanadu is a Canadian company, so let's keep Canadian spelling throughout tutorials.

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# The quantum embedding circuit will consist of two parts.
+# The first one, trainable, will be a parametrised general state preparation
+# scheme :math:`W_a`, with parameters :math:`a`. 
+# In the second one we input the data, and call it :math:`S(x)`.
```
```suggestion
# In the second one we input the data, denoted by :math:`S(x)`.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+    for (i, wire) in enumerate(wires):
+        qml.RZ(thetas[i] * x, wires = [wire])
+###############################################################################
+# By setting the ``thetas`` properly, we achieve the integer-valued spectrum:
```
It was not clear to me why we need an integer valued spectrum. Is there an ultimate reason for it, or just a matter of taste?

emgilfuster replied:
in Fourier analysis we approximate functions f(x) with linear combinations of {sin(nwx), cos(nwx)}, for some fundamental frequency w and integer valued n. Without the condition that n is integer, {sin(nwx), cos(nwx)} don't form an orthonormal basis, and then Fourier theory doesn't hold.
It's a bit weird, we could construct kernels with arbitrary frequencies, but if we want to do this thing of learning a function by matching its Fourier spectrum, then we need to restrict the frequencies to be integer-valued. 

Said otherwise, we have a mechanism that gives us a frequency spectrum and a set of coefficients for any given function. With this quantum kernel first we try to match the frequency spectrum (by fixing the thetas), and then we try to match the coefficients (with the array of amplitudes lower down). The fact we want our spectrum to be integer valued comes from the fact we've chosen the FFT as our mechanism.

I'm not sure I'd know how to convey this in a condensed and non-technical way here. Do you think it's coming too much out of nowhere?

alvaro-at-xanadu replied:
Ok, I see, this comes from $\omega = 1$. Maybe just add: "as required by the Fourier series expansion of a function of period $2\pi$".

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# Finding the solution
+# -------------------------
+#
+# In order to use Newton's method we need the Jacobian of :math:`F_s`.
```
It would be nice to write the Jacobian explicitly here. You wrote it in code, but if someone wants to verify that this is indeed the case, it would be better for them to actually see the formula.

emgilfuster replied:
hm I'm not sure about that one, it's a bit of a big matrix. It's not complicated, but might look scary from the outset? (it's from an old iteration, these w what we call p now)
![image](https://user-images.githubusercontent.com/44003229/149978822-bdbee46d-9397-4e75-bd7b-0861cc137826.png)

Of course I value your zoomed-out perspective more than mine here, just wanted to make sure you don't think this matrix would scare more than it helps.

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# Me neither!
+# But all those probabilities being close to :math:`0` should make us fear some
+# of them must've turned negative.
+# This isn't necessarily bad in general, only for us.
```
So it wasn't a problem in the end? There seems to be no further comment on how we went around this.

emgilfuster replied:
hm yeah that sentence is a bit misleading maybe.
I change it into "This would be fatal for us" and add one sentence on this.
Does it sound better now?

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+Forget about advantages, supremacies, or speed-ups.
+Let us understand better what we can and cannot do with a quantum computer.
+More specifically, in this demo we want to look into quantum kernels and ask
```
```suggestion
More specifically, in this demo, we want to look into quantum kernels and ask
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+Yet, by studying what classes of functions we can realize with quantum kernels,
+we can gain some insight into their inner workings.
+
+Usually, in quantum machine learning (QML) we use parametrized quantum circuits
```
```suggestion
Usually, in quantum machine learning (QML), we use parametrized quantum circuits
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+Usually, in quantum machine learning (QML) we use parametrized quantum circuits
+(PQCs) to find good functions, whatever *good* means here.
+Consequently, since kernels are just one specific kind of well-defined
```
```suggestion
Since kernels are just one specific kind of well-defined
```

alvaro-at-xanadu replied:
Let's start trimming down words to not be unnecessarily wordy. Always a good exercise!

emgilfuster replied:
Thank you for this, I've been battling with this demo for too long and have become unaware of finding these superfluous words :(

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+Consequently, since kernels are just one specific kind of well-defined
+functions, the task of finding a quantum kernel (QK) that approximates a given
+classical one could be posed as an optimisation problem.
+The way to attack this task would be to define a loss function directly
```
```suggestion
The way to attack this task is to define a loss function
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+lens of the Fourier representation of a kernel, which is a common tool in
+classical ML.
+Two functions can only have the same Fourier spectrum if they are the same
+function, and it turns out that for certain classes of quantum circuits `we can
```
```suggestion
function. It turns out that for certain classes of quantum circuits `we can
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+theoretically describe the Fourier spectrum rather well
+<https://pennylane.ai/qml/demos/tutorial_expressivity_fourier_series.html>`_.
+
+Using this theory together with some good old-fashioned convex optimisation, we
```
```suggestion
Using this theory together with some good old-fashioned convex optimization, we
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+Using this theory together with some good old-fashioned convex optimisation, we
+will derive a quantum circuit that approximates the famous Gaussian kernel.
+
+In order to keep the demo short and sweet, we focus on one simple example, but
```
```suggestion
In order to keep the demo short and sweet, we focus on one simple example. The```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+will derive a quantum circuit that approximates the famous Gaussian kernel.
+
+In order to keep the demo short and sweet, we focus on one simple example, but
+the same ideas would apply for more general scenarios.
```
```suggestion
same ideas apply to more general scenarios.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+functions, the task of finding a quantum kernel (QK) that approximates a given
+classical one could be posed as an optimisation problem.
+The way to attack this task would be to define a loss function directly
+quantifying the distance between both functions (the classical kernel function,
```
```suggestion
quantifying the distance between both functions (the classical kernel function
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+----------------------------------------------
+
+We will not be reviewing all the notions of kernels in-depth here.
+Instead, we only need to know that there's an entire branch of ML which
```
```suggestion
Instead, we only need to know that an entire branch of ML which
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+In order to build intuition, we will instead study the link between classical and quantum kernels through the
+lens of the Fourier representation of a kernel, which is a common tool in
+classical ML.
```
```suggestion
classical machine learning.
```

alvaro-at-xanadu replied:
Haven't defined ML acronym.

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+We will not be reviewing all the notions of kernels in-depth here.
+Instead, we only need to know that there's an entire branch of ML which
+revolves around some functions we call kernels.
+If you'd like to see a more comprehensive introduction to where these functions
```
```suggestion
If you'd like to learn more about where these functions
```

alvaro-at-xanadu replied:
More trimming down. Important when sentences are too long :)

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+revolves around some functions we call kernels.
+If you'd like to see a more comprehensive introduction to where these functions
+come from, why they're important, and how we can use them (e.g. with
+PennyLane), you could check out the following demos, which cover different
```
```suggestion
PennyLane), check out the following demos, which cover different
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+For the purposes of this demo, a *kernel* is a real-valued function of two
+variables :math:`k(x_1,x_2)` from a given data domain :math:`x_1,
+x_2\in\mathcal{X}`.
+Most often, and in this demo, we'll deal with real vector spaces as the data
```
```suggestion
In this demo, we'll deal with real vector spaces as the data
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+A kernel has to be symmetric with respect to exchanging both variables
+:math:`k(x_1,x_2) = k(x_2,x_1)`.
+We also enforce kernels to be positive semi-definite, but let's avoid getting
+lost in mathematical lingo, you can trust that all kernels featuring in this
```
```suggestion
lost in mathematical lingo. You can trust that all kernels featuring in this
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+Warm up: Implementing the Gaussian kernel
+-------------------------------------------
+
+First, let's introduce a simple classical kernel that we will try to
```
```suggestion
First, let's introduce a simple classical kernel that we will
```

alvaro-at-xanadu replied:
Sounds more confident 💪

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+plt.show();
+
+###############################################################################
+# In this demo we will consider only this one example, but the arguments we
```
```suggestion
# In this demo, we will consider only this one example. However, the arguments we
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+##################################################################################
+# In practice, we would construct several periodic extensions of the aperiodic
+# function, with increasing periods.
+# This way, we would be able to study the behavior when the period approaches
```
```suggestion
# This way, we can study the behaviour when the period approaches
```

alvaro-at-xanadu replied:
Trimming down! "Behaviour" is the Canadian variant of the spelling (confusing since it's a mix of US and UK spellings 😖 )

emgilfuster replied:
thanks! oh my, I had honestly never heard about Canadiant spelling 😅 thanks for taking the time to check that!

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+    return np.real(coefficients(Gauss_p, 1, d-1)[:d])
+
+##################################################################################
+# With this we are restricted to considering only a finite number of Fourier
```
```suggestion
# We are restricted to considering only a finite number of Fourier
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# coefficients converge to :math:`0` at some point, and we will only need to 
+# consider terms up to this point.
+# Let's look at the coefficients we obtain by setting a low value for the
+# number of coefficients, and then letting it slowly grow:
```
```suggestion
# number of coefficients and then slowly letting it grow:
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# For very small coefficient counts, like :math:`2` and :math:`3`, we see that
+# the last allowed coefficient is still far from :math:`0`.
+# That's a very clear indicator that we need to consider more frequencies.
+# At the same time, it seems like starting :math:`5` or :math:`6` all the
```
```suggestion
# At the same time, it seems like starting at :math:`5` or :math:`6` all the
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# That's a very clear indicator that we need to consider more frequencies.
+# At the same time, it seems like starting :math:`5` or :math:`6` all the
+# non-zero contributions have already been well captured.
+# This is important for us, since it tells us how many qubits we should use at
```
```suggestion
# This is important for us, since it tells us the minimum number of qubits we should use.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# At the same time, it seems like starting :math:`5` or :math:`6` all the
+# non-zero contributions have already been well captured.
+# This is important for us, since it tells us how many qubits we should use at
+# least.
```
```suggestion

```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# least.
+# One can see that every new qubit doubles the number of frequencies we can
+# use, so for :math:`n` qubits, we will have :math:`2^n`.
+# So, at least :math:`6` frequencies means at least :math:`3` qubits, which
```
```suggestion
# At minimum of :math:`6` frequencies means at least :math:`3` qubits, corresponding
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# One can see that every new qubit doubles the number of frequencies we can
+# use, so for :math:`n` qubits, we will have :math:`2^n`.
+# So, at least :math:`6` frequencies means at least :math:`3` qubits, which
+# would correspond to :math:`2^3=8` frequencies.
```
```suggestion
# to :math:`2^3=8` frequencies.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+    return [2 ** i for i in range(n_wires-1, -1, -1)]
+
+###############################################################################
+# Next we introduce the only trainable gate we neet to make use of.
```
```suggestion
# Next, we introduce the only trainable gate we need to make use of.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+#   S(x_2)W_a\lvert0\rangle\rvert^2 \\
+#   &= \lvert\langle0\rvert W_a^\dagger S(x_2-x_1) W_a\lvert0\rangle\rvert^2.
+#
+# In the code, we call :math:`a` ``amplitudes``:
```
```suggestion
# In the code below, the variable ``amplitudes`` corresponds to our set  :math:`a`.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+#   k_a(\delta) = \lvert\langle0\rvert W_a^\dagger
+#   S(\delta)W_a\lvert0\rangle\rvert^2.
+#
+# So far we only wrote the gate layout for the quantum circuit, no measurement!
```
```suggestion
# So far, we only wrote the gate layout for the quantum circuit, no measurement!
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# Computing the QK function on a quantum device
+# -------------------------------------------------
+#
+# Also, at this point we need to set the number of qubits of our computer.
```
```suggestion
# Also, at this point, we need to set the number of qubits of our computer.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+dev = qml.device("default.qubit", wires = n_wires, shots = None)
+
+###############################################################################
+# Next we construct the quantum node:
```
```suggestion
# Next, we construct the quantum node:
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+    return qml.probs(wires = range(n_wires))
+
+###############################################################################
+# Recall that the outputs of QKs can be defined as the probability of obtaining
```
```suggestion
# Recall that the output of QK is defined as the probability of obtaining
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+###############################################################################
+# Recall that the outputs of QKs can be defined as the probability of obtaining
+# the outcome :math:`\lvert0\rangle` when measuring on the computational basis.
```
```suggestion
# the outcome :math:`\lvert0\rangle` when measuring in the computational basis.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+###############################################################################
+# As a couple quality-of-life improvements, we write a function that implements
+# the QK with the lag :math:`\delta` as argument, and one that implements it
```
```suggestion
# the QK with the lag :math:`\delta` as its argument, and one that implements it
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+    return QK_circuit(x1, x2, thetas, amplitudes)[0]
+
+###############################################################################
+# As a couple quality-of-life improvements, we write a function that implements
```
```suggestion
# As a couple of quality-of-life improvements, we write a function that implements
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+###############################################################################
+# One can see that the stationary kernel with this particular initial state has
+# a decaying spectrum which looks similar to :math:`1/\lvert x\rvert` - but not
```
```suggestion
# a decaying spectrum that looks similar to :math:`1/\lvert x\rvert` - but not
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# Fourier spectrum, our job would be done here.
+# However, the equations derived in the literature are not trivial to solve.
+#
+# As mentioned in the introduction, one could just "learn" this relation: take
```
As mentioned in the introduction, one could just “learn” this relation, that is, tune the parameters of the quantum kernel in a gradient-based manner until it matches the classical one.

alvaro-at-xanadu replied:
This sentence is too long and convoluted. Suggested change above ⏫

emgilfuster replied:
good, thanks!

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# gradient-based manner until it matches the classical one.
+#
+# We want to take an intermediate route between analytical solution and
+# black-box optimisation here.
```
```suggestion
# black-box optimization here.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# We want to take an intermediate route between analytical solution and
+# black-box optimisation here.
+# For that, we derive an equation that links the amplitudes to the spectrum we
+# want to construct, and then use old-fashioned convex optimisation to find the
```
```suggestion
# want to construct and then use old-fashioned convex optimization to find the
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# want to construct, and then use old-fashioned convex optimisation to find the
+# solution.
+# If you are not interested in the details, you can just jump to the last plots
+# of this demo and confirm that it allows us to emulate the Gaussian kernel
```
```suggestion
# of this demo and confirm that we can to emulate the Gaussian kernel
```

alvaro-at-xanadu replied:
Not clear what 'it' refers to.

emgilfuster replied:
agree, thanks!

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+###############################################################################
+# These closed-form equations allow us to find the solution numerically, using
+# Newton's method!
+# Newton's method is a classical one from convex optimisation theory.
```
```suggestion
# Newton's method is a classical one from convex optimization theory.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+###############################################################################
+# Showing that this is indeed :math:`\nabla F_s` is left as an exercise for the
+# reader.
+# For Newton's method we also need an initial guess.
```
```suggestion
# For Newton's method, we also need an initial guess.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# Showing that this is indeed :math:`\nabla F_s` is left as an exercise for the
+# reader.
+# For Newton's method we also need an initial guess.
+# Finding a good initial guess requires some tinkering, different problems will
```
```suggestion
# Finding a good initial guess requires some tinkering; different problems will
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# But all those probabilities being close to :math:`0` should make us fear some
+# of them must've turned negative.
+# This isn't necessarily bad in general, only for us.
+# For ``MottonenStatePreparation`` we'll need to give ``amplitudes`` as one of the
```
```suggestion
# For ``MottonenStatePreparation``, we'll need to give ``amplitudes`` as one of the
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+# of them must've turned negative.
+# This isn't necessarily bad in general, only for us.
+# For ``MottonenStatePreparation`` we'll need to give ``amplitudes`` as one of the
+# arguments, which is the component-wise square root of ``probabilities`` so it
```
```suggestion
# arguments, which is the component-wise square root of ``probabilities``, so it
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+
+###############################################################################
+# It seems like it does!
+# But as we just said this is still only the predicted spectrum.
```
```suggestion
# But as we just said, this is still only the predicted spectrum.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,761 @@
+    return np.real(coefficients(lambda x: QK(x, thetas, amplitudes), 1, d-1))
+
+###############################################################################
+# And with this we can finally visualize how the Fourier spectrum of the
```
```suggestion
# And with this, we can finally visualize how the Fourier spectrum of the
```

alvaro-at-xanadu reviewed (commented):
I think this is a well-written and interesting tutorial. I only have a few comments about the content, but it may be that I don't know much about quantum kernels.

I suggested some changes to the grammar, since some sentences were hard to parse. Most corrections are minor. 

Excellent work :)

emgilfuster commented:
Thank you so much for that review @alvaro-at-xanadu!
I appreciate the effort you gave to improving the demo, also the misaligned spelling!

I implemented all the suggestions except for one, see the comment on the jacobian matrix :)
I also tried to explain the necessity for the spectrum to be integer valued, and why it might be difficult to explain in the demo.

Have a nice week

mariaschuld commented:
@alvaro-at-xanadu could you kindly approve when you're happy? We'll then be merge ready!

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,765 @@
+(PQCs) to find good functions, whatever *good* means here.
+Since kernels are just one specific kind of well-defined
+functions, the task of finding a quantum kernel (QK) that approximates a given
+classical one could be posed as an optimisation problem.
```
```suggestion
classical one could be posed as an optimization problem.
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,765 @@
+# *What can we do?!*
+#
+# We can cook up a periodic extension to the Gaussian kernel, for a given
+# period :math:`2L` (we take :math:`2\pi` as default):
```
```suggestion
# period :math:`2L` (we take :math:`L=\pi` as default):
```

alvaro-at-xanadu commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,765 @@
+
+###############################################################################
+# One can see that the stationary kernel with this particular initial state has
+# a decaying spectrum that looks similar to :math:`1/\lvert x\rvert` - but not
```
```suggestion
# a decaying spectrum that looks similar to :math:`1/\lvert x\rvert` — but not
```

alvaro-at-xanadu replied:
Changed to em dash

emgilfuster commented:
Implemented the last changes, thanks for the new catches!

alvaro-at-xanadu reviewed (approved):
We're good to go!

peterse commented on demonstrations/tutorial_classical_kernels.py:
```diff
@@ -0,0 +1,765 @@
+# Some math shows that setting :math:`\vartheta_i=2^{n-i}`, for
+# :math:`\{1,\ldots,n\}` produces the desired outcome.
+
+def make_thetas(m_wires):
```
```suggestion
def make_thetas(n_wires):
```

emgilfuster commented:
Hi hi!

I couldn't resist adding the 300th comment on this discussion :)
Is everything release-ready? I guessed this was put on hold because of qhack ^^

mariaschuld commented:
@emgilfuster sorry, I should have followed up on this. @CatalinaAlbornoz, would it be possible to merge this soon? :)
